File: OC2.c


#include <stdio.h> // Включаем стандартную библиотеку ввода/вывода
#include <stdlib.h> // Включаем стандартную библиотеку для выделения памяти и функций выхода
#include <unistd.h> // Включаем POSIX API для системных вызовов
#include <signal.h> // Подключаем библиотеку обработки сигналов
#include <arpa/inet.h> // Включаем библиотеку для манипулирования интернет-адресами
#include <sys/socket.h> // Подключаем библиотеку для операций с сокетами
#include <sys/types.h> // Подключаем библиотеку для типов данных
#include <errno.h> // Подключаем библиотеку для обработки ошибок

volatile sig_atomic_t g_got_sighup = 0; // Объявляем глобальную переменную для хранения флага сигнала SIGHUP
int g_accepted_socket = -1; // Объявляем глобальную переменную для хранения принятого дескриптора сокета

void handle_signal(int in_signum) { // Определяем функцию для обработки сигналов
    if (in_signum == SIGHUP) { // Если сигнал SIGHUP
        g_got_sighup = 1;// Устанавливаем флаг в 1
    } else { 
        printf("Received signal %d\n", in_signum); // Распечатываем номер сигнала
    }
}

int main() { // Определяем главную функцию
    struct sigaction signal_action; // Объявляем структуру для хранения действия сигнала
    signal_action.sa_handler = handle_signal;// Устанавливаем функцию-обработчик
    signal_action.sa_flags = SA_RESTART;// Устанавливаем флаг перезапуска системного вызова после обработки сигнала
    sigaction(SIGHUP, &signal_action, NULL); // Регистрируем действие сигнала для SIGHUP

    // Блок создания сокета
    int server_socket, client_socket; // Объявляем две переменные для хранения дескрипторов сокетов сервера и клиента
    struct sockaddr_in server_addr, client_addr; // Объявляем две структуры для хранения информации об адресах сервера и клиента
    socklen_t client_len = sizeof(client_addr); // Объявляем переменную для хранения размера структуры адреса клиента

    if ((server_socket = socket(AF_INET, SOCK_STREAM, 0)) == -1) { // Создаем сокет для сервера с использованием протоколов IPv4 и TCP и проверяем наличие ошибок
        perror("Error creating socket");// Распечатываем сообщение об ошибке
        exit(EXIT_FAILURE); // Выходим из программы со статусом сбоя
    }

    server_addr.sin_family = AF_INET; // Устанавливаем семейство адресов IPv4
    server_addr.sin_addr.s_addr = INADDR_ANY; // Устанавливаем любой IP-адрес
    server_addr.sin_port = htons(8080); // Устанавливаем номер порта 8080 в сетевом порядке байтов

    if (bind(server_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) == -1) { // Привязываем сокет к адресу сервера и проверяем наличие ошибок
        perror("Error binding socket"); 
        exit(EXIT_FAILURE); // Выходим из программы со статусом сбоя
    }

    if (listen(server_socket, 1) == -1) { // Прослушиваем входящие соединения на сокете и проверяем наличие ошибок
        perror("Error listening for connections"); 
        exit(EXIT_FAILURE); // Выходим из программы со статусом сбоя
    }

    printf("Server listening on port 8080\n"); // Распечатываем сообщение, указывающее, что сервер прослушивает

    char buffer[1024]; // Объявляем буфер для хранения полученных данных

    fd_set read_fds; // Объявляем набор дескрипторов файлов для хранения сокетов, которые будут отслеживаться на чтение
    FD_ZERO(&read_fds); // Инициализируем набор равным нулю
    // Блок обработки
    while (1) { 
        if (g_got_sighup) { // Если установлен флаг SIGHUP
            printf("Received SIGHUP signal. Closing the server.\n"); // Распечатываем сообщение, указывающее, что сервер закрывается
            if (g_accepted_socket != -1) {// Если есть принятый сокет
                close(g_accepted_socket); // Закрываем сокет
            }
            exit(0); // Выходим из программы со статусом успеха
        }

FD_SET(server_socket, &read_fds); // Добавляем сокет сервера в набор файловых дескрипторов

        struct timespec timeout; // Объявляем структуру для хранения значения таймаута
        timeout.tv_sec = 1; // Устанавливаем таймаут на 1 секунду
        timeout.tv_nsec = 0; // Устанавливаем наносекунды в 0
        
        sigset_t mask; // Объявляем набор сигналов для хранения сигналов для блокировки
    // Вызов функции pselect, чтобы отслеживать набор файловых дескрипторов для чтения и сохранять результа
        int result = pselect(serversocket + 1, &readfds, NULL, NULL, &timeout, &mask); 
        
        if (result == -1) { // Если результат равен -1, это указывает на ошибку
            if (errno == EINTR) { // Если ошибка вызвана прерванным системным вызовом
                printf("'pselect' was interrupted by a signal.\n"); // pselect был прерван
                continue;
            } else {
                perror("Error in pselect");
                break;// Разрыв цикла
            }
        }

        if (FD_ISSET(server_socket, &read_fds)) {// Если сокет сервера готов к чтению, что указывает на входящее соединение
            // принимаем соединение и сохраняем дескриптор сокета клиента, а также проверяем наличие ошибок
            if ((client_socket = accept(server_socket, (struct sockaddr*)&client_addr, &client_len)) == -1) {
                perror("Error accepting connection");
                continue;
            }
            // Распечатываем адрес клиента и номер порта
            printf("Accepted connection from %s:%d\n", inet_ntoa(client_addr.sin_addr), ntohs(client_addr.sin_port));

            if (g_accepted_socket == -1) {// Если принятого сокета нет
                g_accepted_socket = client_socket;// Устанавливаем глобальную переменную в клиентский сокет
            } else {
                close(client_socket);// Закрываем клиентский сокет
            }
        }
    // Получаем данные из принятого сокета и сохраняем количество полученных байтов
        size_t bytes_received = recv(g_accepted_socket, buffer, sizeof(buffer), 0);
        if (bytes_received > 0) { // Если количество полученных байт положительное
            buffer[bytes_received] = '\0'; // Добавляем нулевой терминатор в буфер
            printf("Received %ld bytes: %s\n", bytes_received, buffer);// Выводим количество байт и полученные данные
        } else if (bytes_received == 0) { // Если количество полученных байтов равно нулю, это означает, что соединение закрыто
            printf("Connection closed by client.\n");
            close(g_accepted_socket);// Закрываем принятый сокет
            g_accepted_socket = -1; // Устанавливаем глобальную переменную в -1
        } 
        else { // В противном случае, что указывает на ошибку
            perror("Error receiving data");
        }
    }
    return 0; // Возвращаем 0, чтобы указать на успех
}
         

File: OC2.py

import socket
import signal
import threading
import sys

# Мьютекс
m_mutex = threading.Lock()
# Уведомление для ожидающиъ потоков о изменении состояния
m_notification = threading.Condition()
# Обработчик сигнала (флаг)
wasSignalHangUp = 0

# Обработка сигнала
def SignalHangUpHandler(signal, frame):
    # Когда сигнал будет получен, установим обработчику значение 1
    global wasSignalHangUp
    wasSignalHangUp = 1

# Обработка данных (полученных от клиента через сокет)
def ConnectionManagment(clientSocket):
    # Создаём массив для хранения данных, полученных от клиента
    buffer = bytearray(1024)
    try:
        # Считываем данные из сокета в буфер
        bytesReceived = clientSocket.recv_into(buffer)
        # Проверяем, были ли фактически получены какие либо данные от клиента
        if bytesReceived > 0:
            # Захватываем мьютекс перед выводом в консоль
            with m_mutex:
                # Если данные были получены, выводим количество байт
                print(f"Полученные данные от клиента: {bytesReceived} байт")
    except socket.error as e:
        # Выводим сообщение об ошибке при чтении из сокета
        print(f"Ошибка при чтении из сокета: {e}", file=sys.stderr)

# Принятие новых соединений от клиентов
def AcceptConnections(serverSocket):
    try:
        # Создаём структуру для хранения информации о клиенте, подключившемся к серверу
        clientSocket, clientAddress = serverSocket.accept()
        # Проверяем, успешно ли был создан сокет для нового клиента
        if clientSocket:
            with m_mutex:
                # Если сокет успешно создан, выводится сообщение о том, что новое соединение было принято
                print("Новое соединение принято")
                # Выводим IP адрес и порт клиента
                print(f"IP адрес клиента: {clientAddress[0]}")
                print(f"Порт клиента: {clientAddress[1]}")
            # Создаётся новый поток, который будет управлять соединением с клиентом
            connectionThread = threading.Thread(target=ConnectionManagment, args=(clientSocket,))
            # Поток запускается независимо от главного потока и работает параллельно с ним
            connectionThread.start()
            # Сокет закрывается, так как он больше не нужен для принятия новых соединений
            clientSocket.close()
        else:
            # Вывод сообщения об ошибке в случае неудачной попытки принятия соединения
            print("Ошибка при попытке принять соединение", file=sys.stderr)
    except socket.error as e:
        # Выводим сообщение об ошибке при принятии соединения
        print(f"Ошибка при принятии соединения: {e}", file=sys.stderr)


def main():
    # Запускаем в работу сокеты
    result = socket.has_ipv6
    # Проверяем ошибки при инициализации библиотеки сокетов
    if not result:
        # Выводим сообщение об ошибке
        print("Не удалось выполнить WSAStartup", file=sys.stderr)
        return 1
    # Создаём серверный сокет
    serverSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    # Проверяем ошибки при создании серверного сокета
    if serverSocket == -1:
        # Выводим сообщение об ошибке
        print("Не удалось создать серверный сокет", file=sys.stderr)
        return 1
    # Задаём параметры серверного сокета
    serverAddress = ("", 15200) # Адрес и порт для прослушивания
    try:
        # Связываем серверный сокет с адресом и портом
        serverSocket.bind(serverAddress)
        # Переводим серверный сокет в режим прослушивания
        serverSocket.listen(socket.SOMAXCONN)
    except socket.error as e:
        # Выводим сообщение об ошибке при связывании или прослушивании сокета
        print(f"Ошибка при связывании или прослушивании сокета: {e}", file=sys.stderr)
        return 1
    # Устанавливаем обработчик сигнала
   # signal.signal(signal.SIGINT, SignalHangUpHandler)
    # Выводим сообщение о том, что сервер запущен и ожидает подключений
    print("Сервер запущен и ожидает подключений")
    # Запускаем бесконечный цикл для принятия новых соединений
    while True:
        # Проверяем, был ли получен сигнал
        if wasSignalHangUp == 1:
            # Если сигнал был получен, выходим из цикла
            break
        # Принимаем новые соединения от клиентов
        AcceptConnections(serverSocket)
    # Закрываем серверный сокет
    serverSocket.close()
    return 0

if __name__ == "__main__":
    main()